### PR TITLE
Property support

### PR DESCRIPTION
**Pull request by jasonwaters**

### CODE REVIEW

File: packages/target-decisioning-engine/src/decisionProvider.js [hunk 1]
```diff
@@ -16,6 +17,7 @@ import {
 } from "./postProcessors";
 import { ruleEvaluator } from "./ruleEvaluator";
 import { LOG_PREFIX } from "./constants";
+import { byPropertyToken } from "./filters";
 
 const LOG_PREAMBLE = `${LOG_PREFIX}.DecisionProvider`;
```
Comment (XDex) [inline packages/target-decisioning-engine/src/decisionProvider.js]: we usually call it just `LOG_TAG`

File: packages/target-decisioning-engine/src/filters.js [hunk 2]
```diff
@@ -0,0 +1,17 @@
+/* eslint-disable import/prefer-default-export */
+import { isUndefined } from "@adobe/target-tools";
+
+export function byPropertyToken(propertyToken) {
+  /**
+   * @param {import("../types/DecisioningArtifact").Rule} rule
+   * @return {boolean}
+   */
+  function filter(rule) {
```
Comment (XDex) [inline packages/target-decisioning-engine/src/filters.js:14]:
```suggestion
  function filter(rule = {}) {
```

File: packages/target-decisioning-engine/src/filters.js [hunk 3]
```diff
@@ -0,0 +1,17 @@
...
+    const { propertyTokens = [] } = rule;
+
+    return isUndefined(propertyToken)
+      ? propertyTokens.length === 0
```
Comment (XDex) [inline packages/target-decisioning-engine/src/filters.js:18]:
```suggestion
      ? !propertyTokens.length
```

File: packages/target-decisioning-engine/src/filters.js [hunk 4]
```diff
@@ -0,0 +1,17 @@
...
+      : propertyTokens.length === 0 || propertyTokens.includes(propertyToken);
```
Comment (XDex) [inline packages/target-decisioning-engine/src/filters.js:19]:
```suggestion
      : !propertyTokens.length || propertyTokens.includes(propertyToken);
```

File: packages/target-decisioning-engine/src/utils.js [hunk 5]
```diff
@@ -111,3 +118,67 @@ export function cloneDeep(obj) {
...
+  return CDN_BASE[env];
+}
+
+/**
+ *
+ * @param {String} environmentName
+ * @param logger
+ */
+export function getValidEnvironment(environmentName, logger) {
+  const isValid = POSSIBLE_ENVIRONMENTS.includes(environmentName);
```
Comment (XDex) [inline packages/target-decisioning-engine/src/utils.js:137]:
```suggestion
  const isInvalid = !POSSIBLE_ENVIRONMENTS.includes(environmentName);
```

File: packages/target-decisioning-engine/src/utils.js [hunk 6]
```diff
@@ -111,3 +118,67 @@ export function cloneDeep(obj) {
...
+
+  if (!isValid) {
```
Comment (XDex) [inline packages/target-decisioning-engine/src/utils.js:139]:
```suggestion
  if (isInvalid) {
```

File: packages/target-decisioning-engine/src/utils.js [hunk 7]
```diff
@@ -111,3 +118,67 @@ export function cloneDeep(obj) {
...
+    getLogger(logger).debug(
+      Messages.INVALID_ENVIRONMENT(environmentName, ENVIRONMENT_PROD)
+    );
```
Comment (XDex) [inline packages/target-decisioning-engine/src/utils.js:142]:
```suggestion
    );
return ENVIRONMENT_PROD;
```

File: packages/target-decisioning-engine/src/utils.js [hunk 8]
```diff
@@ -111,3 +118,67 @@ export function cloneDeep(obj) {
...
+  }
+
+  return isValid ? environmentName : ENVIRONMENT_PROD;
```
Comment (XDex) [inline packages/target-decisioning-engine/src/utils.js:145]:
```suggestion
  return environmentName;
```

File: packages/target-decisioning-engine/src/utils.js [hunk 9]
```diff
@@ -111,3 +118,67 @@ export function cloneDeep(obj) {
...
+
+  return [
+    getCdnBasePath(cdnEnvironment),
+    client,
+    targetEnvironment,
+    `v${SUPPORTED_ARTIFACT_MAJOR_VERSION}`,
+    propertyToken,
+    "rules.json"
+  ]
+    .filter(value => !isUndefined(value))
```
Comment (XDex) [inline packages/target-decisioning-engine/src/utils.js]:
```suggestion
    .filter(value => isDefined(value))
```

File: packages/target-nodejs-sdk/demo/demo.js [hunk 10]
```diff
@@ -55,25 +52,24 @@ function getOffers() {
...
-      : {}
-  )
-);
+  fetchApi: getFetchWithProxy(),
+  executionMode: "local",
+  // environment: "staging",
+  // cdnEnvironment: "staging",
+  clientReadyCallback: getOffers,
+  logger: {
+    // eslint-disable-next-line no-console
```
Comment (XDex) [inline packages/target-nodejs-sdk/demo/demo.js]: I think you can set this exception at the top of the file just once for the entire file.

File: packages/target-nodejs-sdk/demo/package.json [hunk 11]
```diff
@@ -0,0 +1,16 @@
+{
+  "name": "demo",
```
Comment (XDex) [inline packages/target-nodejs-sdk/demo/package.json]:
```suggestion
  "name": "target-nodejs-sdk-demo",
```

File: packages/target-nodejs-sdk/src/index.js [hunk 12]
```diff
@@ -100,8 +103,11 @@ export default function bootstrap(fetchApi) {
      * @param {Number} options.pollingInterval (Local Decisioning) Polling interval in ms, default: 30000
      * @param {String} options.artifactLocation (Local Decisioning) Fully qualified url to the location of the artifact, optional
      * @param {String} options.artifactPayload (Local Decisioning) A pre-fetched artifact, optional
+     * @param {Number} options.environment The target environment name, defaults to production, optional
```
Comment (XDex) [inline packages/target-nodejs-sdk/src/index.js]:
```suggestion
     * @param {Number} options.environment The Target environment name, defaults to production, optional
```

File: packages/target-tools/src/utils.js [hunk 13]
```diff
@@ -110,3 +112,33 @@ export function objectWithoutUndefinedValues(obj) {
...
+  return token;
+}
+
+export function getProperty(config = {}, request = {}, logger) {
+  const configPropertyToken = config.propertyToken;
+  const requestPropertyToken = getPropertyToken(request.property);
+  const propertyToken = requestPropertyToken || configPropertyToken;
+
+  if (
+    !isUndefined(requestPropertyToken) &&
```
Comment (XDex) [inline packages/target-tools/src/utils.js]:
```suggestion
    isDefined(requestPropertyToken) &&
```

**Review by XDex — APPROVED** [review]
Approved with some changes requested

File: packages/target-decisioning-engine/src/filters.js (hunk 3 above)
Comment (jasonwaters) [reply packages/target-decisioning-engine/src/filters.js:18]: I prefer `propertyTokens.length === 0` because I think it is more readable than `!propertyTokens.length`

**Comment by jasonwaters** [comment]
Thanks for your review @XDex !

File: packages/target-nodejs-sdk/demo/package.json [hunk 14]
```diff
@@ -0,0 +1,16 @@
...
+  "name": "target-nodejs-sdk-demo",
+  "version": "1.0.0",
+  "description": "",
+  "main": "demo.js",
+  "scripts": {
+    "start": "node demo.js"
+  },
+  "keywords": [],
+  "author": "",
+  "license": "ISC",
```
Comment (5amfung) [inline packages/target-nodejs-sdk/demo/package.json]: Should license be `Apache-2.0` because LICENSE file indicates it's Apache 2?

File: packages/target-decisioning-engine/src/artifactProvider.spec.js [hunk 15]
```diff
@@ -280,3 +287,67 @@ describe("artifactProvider", () => {
...
+
+describe("determineArtifactLocation", () => {
+  it("CDN host environment can be overridden", () => {
+    expect(
+      determineArtifactLocation({
+        client: "someClientId",
+        cdnEnvironment: "staging"
+      })
+    ).toEqual(
+      `${CDN_BASE_STAGE}/someClientId/production/v${SUPPORTED_ARTIFACT_MAJOR_VERSION}/rules.json`
```
Comment (5amfung) [inline packages/target-decisioning-engine/src/artifactProvider.spec.js:299]: We might need a test for version which is an empty space, like the following:

```
.../{client}/{env}/rules.json
```

**Review by 5amfung — APPROVED** [review]
LGTM.  Just 2 minor questions below.

File: packages/target-decisioning-engine/src/artifactProvider.spec.js (hunk 15 above)
Comment (jasonwaters) [reply packages/target-decisioning-engine/src/artifactProvider.spec.js:299]: > We might need a test for version which is an empty space, like the following:
> 
> ```
> .../{client}/{env}/rules.json
> ```

Actually, starting with this version of the SDK, a version is always required in the URL.  The reason we keep publishing artifacts without version is to support prior releases of the alpha.

File: packages/target-nodejs-sdk/demo/package.json (hunk 14 above)
Comment (jasonwaters) [reply packages/target-nodejs-sdk/demo/package.json]: Good catch.  Looks like I missed that in some of the other package.json files as well.  Thanks!
